Introduces a speechgenerator pool

zorkow opened the pull request:
PR introduces a `GeneratorPool` data structure that 
* coordinates the computation of speech. 
* takes care of async locale loading.

This makes the speech computation considerably cleaner. In particular there is no longer speech computation in the semantic enrichment method. Speech can be computed if the `speech="deep"` option is set. 
All Speech is now computed in the `attachSpeech` method, where originally only the `aria-label` was set. 

Since the `attachSpeech` state is called after rerender, this also takes care of computing speech for the collapsed elements etc. 

Another side effect is that we no longer use the walkers provided by SRE, which makes syncing a lot easier. At the moment only the walking and summary functionality is ported. The rest will be ported in a subsequent PR. 

**Note:** Currently only works with SRE branch `new_explorer_changes`. I'll try to merge that asap.

zorkow commented:
@dpvc I've made the changes as discussed today. PTAL.

dpvc commented on ts/a11y/explorer/ExplorerPool.ts:
```diff
@@ -295,6 +297,16 @@ export class ExplorerPool {
+   *
+   * @return {SpeechExplorer}
+   */
+  public speech(): SpeechExplorer {
```
Would this be better as a getter so you don't need the `()` every time?  I always go back and forth about that.

zorkow replied:
Done.

dpvc commented on ts/a11y/speech/GeneratorPool.ts:
```diff
@@ -0,0 +1,353 @@
+
+  private _update(options: OptionList) {
+    let update = false;
+    if (options?.sre?.braille !== this.currentBraille) {
+      this.currentBraille = options.sre.braille;
```
If `options` or `options.sre` don't exist, then the `if` condition will be true, and you will end up with an error from `options.sre.braille` on line 141.  I'm not sure the question marks are needed in the `if` condition, but if you need to be concerned about that, I would suggest something like

``` js
if (!options || !options.sre) return false;
```

at the top.  Similarly for line 145.

dpvc commented on ts/a11y/speech/GeneratorPool.ts:
```diff
@@ -0,0 +1,353 @@
+    );
+  }
+
+  private copyAttributes(xml: Element, node: N, attr: string) {
```
Do you need jsdoc for this function and the `attrList` below?

zorkow replied:
I thought I had caught them all. 
jsdoc added.

dpvc commented on ts/a11y/explorer/KeyExplorer.ts:
```diff
@@ -248,11 +291,61 @@ export class SpeechExplorer extends AbstractExplorer<string> implements KeyExplo
+    return node;
+  }
+
+  private refocus(node: HTMLElement) {
```
Do you need jsdoc for this function and the one below?

zorkow replied:
ditto

dpvc reviewed (changes requested):
I've made some comments about some minor issues.  See what you think.

zorkow commented on ts/a11y/explorer.ts:
```diff
@@ -132,8 +123,11 @@ export function ExplorerMathItemMixin<B extends Constructor<HTMLMATHITEM>>(
+      if (speech.attached && speech.active) {
+        const focus = speech.semanticFocus();
+        this.refocus = focus ? focus.id : null;
+      }
```
`this.explorers` can be empty. I've changed both this and the next method to reflect this.

zorkow commented:
Incorporated the review suggestions.
Also I fixed an issue with the walking when elements are being collapsed. They need to have some `data-semantic-` attributes set explicitly.

PTAL.

zorkow commented on ts/a11y/explorer.ts:
```diff
@@ -145,11 +139,11 @@ export function ExplorerMathItemMixin<B extends Constructor<HTMLMATHITEM>>(
     public updateDocument(document: ExplorerMathDocument) {
       super.updateDocument(document);
-      this.refocus && this.typesetRoot.focus();
+      this.explorers.speech().restarted = this.refocus;
```
Done.

dpvc reviewed (approved):
lgtm
